Upstream changelog:
  fix ancestry checking to not be exponentially slow on branchy ancestry

Backport from revision 911f3ce2c66ade267eec549dc15bcef420a1686d.
Index: monotone-0.47/database.cc
===================================================================
--- monotone-0.47.orig/database.cc	2010-06-05 01:36:37.636176391 +0100
+++ monotone-0.47/database.cc	2010-06-05 01:36:44.581925485 +0100
@@ -2623,6 +2623,7 @@
 
   vector<revision_id> todo;
   todo.push_back(ancestor);
+  set<revision_id> seen;
   while (!todo.empty())
     {
       revision_id anc = todo.back();
@@ -2634,11 +2635,16 @@
         {
           if (*i == child)
             return true;
+          else if (seen.find(*i) != seen.end())
+            continue;
           else
             {
               get_rev_height(*i, anc_height);
               if (child_height > anc_height)
-                todo.push_back(*i);
+                {
+                  seen.insert(*i);
+                  todo.push_back(*i);
+                }
             }
         }
     }
